src/representation_theory/Rep.lean

```diff
@@ -0,0 +1,175 @@
+and this type comes equipped with a `module k V` instance.
+Also `V.ρ` gives the homomorphism `G →* (V →ₗ[k] V)`.
+
+Conversely, given a homomorphism `ρ : G →* (V →ₗ[k] V)`,
+you can construct the bundled representation as `Rep.of ρ`.
+
+We verify that `Rep k G` has all limits and colimits, and is a monoidal category.
+-/
+
+universes u
+
+open category_theory
+open category_theory.limits
+
+/-- The category of `k`-linear representations of a monoid `G`. -/
+@[derive [large_category, concrete_category, has_limits, has_colimits]]
+abbreviation Rep (k G : Type u) [ring k] [monoid G] :=
+Action (Module.{u} k) (Mon.of G)
+
+namespace Rep
+
+variables {k G : Type u} [ring k] [monoid G]
+
+instance : has_coe_to_sort (Rep k G) (Type u) := concrete_category.has_coe_to_sort _
+
+instance (V : Rep k G) : add_comm_group V :=
+by { change add_comm_group ((forget₂ (Rep k G) (Module k)).obj V), apply_instance, }
+
+instance (V : Rep k G) : module k V :=
+by { change module k ((forget₂ (Rep k G) (Module k)).obj V), apply_instance, }
+
+-- This works well with the new design for representations:
+def ρ' (V : Rep k G) : G →* (V →ₗ[k] V) := V.ρ
+
+/-- Lift an unbundled representation to `Rep`. -/
+@[simps]
+def of {V : Type u} [add_comm_group V] [module k V] (ρ : G →* (V →ₗ[k] V)) : Rep k G :=
+⟨Module.of k V, ρ⟩
+
+@[simp] lemma of_ρ' {V : Type u} [add_comm_group V] [module k V] (ρ : G →* (V →ₗ[k] V)) :
+  (of ρ).ρ' = ρ := rfl
+
+-- Verify that limits are calculated correctly.
+noncomputable example : preserves_limits (forget₂ (Rep k G) (Module.{u} k)) :=
+by apply_instance
+noncomputable example : preserves_colimits (forget₂ (Rep k G) (Module.{u} k)) :=
+by apply_instance
+
+end Rep
+
+namespace Rep
+variables {k G : Type u} [comm_ring k] [monoid G]
+
+noncomputable theory
+
+def module_monoid_algebra (V : Rep k G) : module (monoid_algebra k G) V :=
+representation.as_module V.ρ
+
+local attribute [instance] module_monoid_algebra
+
+@[simp]
+lemma turkle {V : Rep k G} (r : monoid_algebra k G) (x : V) :
+  r • x = (monoid_algebra.lift k G _) V.ρ' r x :=
+rfl
+
+lemma algebra_map_smul' {V : Rep k G} {r : k} {v : V} :
+  algebra_map k (monoid_algebra k G) r • v = r • v :=
+begin
+  simp,
+end
+
+local attribute [-instance] Rep.module
+
+lemma quux {V : Rep k G} {g : G} {v : V} : monoid_algebra.of k G g • v = (V.ρ g).to_fun v :=
+begin
+  simp,
+  refl,
+end
+
+def to_Module_monoid_algebra_map {V W : Rep k G} (f : V ⟶ W) :
+  Module.of (monoid_algebra k G) V ⟶ Module.of (monoid_algebra k G) W :=
+{ map_smul' := λ r x, begin
+    dsimp,
+    apply monoid_algebra.induction_on r,
+    { intro g,
+      simp only [one_smul, monoid_algebra.lift_single, monoid_algebra.of_apply],
+      exact congr_hom (f.comm g) x, },
+    { intros g h gw hw, simp only [map_add, add_left_inj, linear_map.add_apply, hw, gw], },
+    { intros r g w,
+      simp only [alg_hom.map_smul, w, ring_hom.id_apply,
+        linear_map.smul_apply, linear_map.map_smulₛₗ], },
+  end,
```

Comment:
I feel like this should already exist as a lemma somewhere. What's the statement with all the category theory stuff unfolded?

Reply:
Good question. The statement was
```
lemma aux {k G : Type*} [comm_ring k] [monoid G]
  (V W : Type*) [add_comm_group V] [add_comm_group W] [module k V] [module k W]
  (ρ : G →* V →ₗ[k] V) (σ : G →* W →ₗ[k] W)
  (f : V →ₗ[k] W) (w : ∀ (g : G), f.comp (ρ g) = (σ g).comp f)
  (r : monoid_algebra k G) (x : V) :
  f ((((monoid_algebra.lift k G (V →ₗ[k] V)) ρ) r) x) =
    (((monoid_algebra.lift k G (W →ₗ[k] W)) σ) r) (f x) :=
```
and proving this separately results in much faster typechecking.